core/src/gl/mesh.cpp

```diff
@@ -143,51 +130,65 @@ void MeshBase::upload() {
         }
 
         // Buffer element index data
-        RenderState::indexBuffer(m_glIndexBuffer);
+        rs.indexBuffer(m_glIndexBuffer);
 
         GL_CHECK(glBufferData(GL_ELEMENT_ARRAY_BUFFER, m_nIndices * sizeof(GLushort), m_glIndexData, m_hint));
 
         delete[] m_glIndexData;
         m_glIndexData = nullptr;
     }
 
-    m_generation = RenderState::generation();
+    m_generation = rs.generation();
+
+    m_disposer.reset(new Disposer(rs, [=](RenderState& rs){
+        // Deleting a index/array buffer being used ends up setting up the current vertex/index buffer to 0
+        // after the driver finishes using it, force the render state to be 0 for vertex/index buffer
+        if (rs.isValidGeneration(m_generation)) {
+            if (m_glVertexBuffer) {
+                rs.vertexBufferUnset(m_glVertexBuffer);
+                GL_CHECK(glDeleteBuffers(1, &m_glVertexBuffer));
+            }
+            if (m_glIndexBuffer) {
+                rs.indexBufferUnset(m_glIndexBuffer);
+                GL_CHECK(glDeleteBuffers(1, &m_glIndexBuffer));
+            }
+            m_vaos.dispose();
+        }
```

Comment:
This is a copy of m_vaos before it was initialized. Since this keeps a reference to RenderState in disposer one could also keep a reference to it in Mesh and create the disposer in Mesh's destructor.